### ISSUE TITLE
UnicodeDecodeError

### ISSUE THREAD

**Issue by doldman** [koenedaele/skosprovider#16]
Just started by using the DictionaryProvider example in the documents. I get this error

File "C:\Python34\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 31946: character maps to <undefined>

**Comment by koenedaele** [comment]
Hi Dominic,

coiuld you post the code you're using?

**Comment by doldman** [comment]
Hi Koen,

yes. I just pasted in the example in the documentation with some imports, After finding your project

from skosprovider.providers import DictionaryProvider
from skosprovider.uri import UriPatternGenerator
from skosprovider.skos import ConceptScheme
from skosprovider.utils import unicode_literals

provider = DictionaryProvider(
   {
        'id': 'TREES',
        'default_language': 'nl',
        'subject': ['biology']
    },
    [larch, chestnut, species],
    uri_generator=UriPatternGenerator('http://id.trees.org/types/%s'),
    concept_scheme=ConceptScheme('http://id.trees.org')
)

**Comment by koenedaele** [comment]
I've added a sample script: https://github.com/koenedaele/skosprovider/blob/master/examples/api.py

Could you try running this?

The problem you're running into is probably Windows specific and has to do with character encodings. The sample I posted has a special header to tell python the data in the file is UTF-8. But I thought that was only needed for python 2.x and you seem to be running 3.4.

If you can confirm the sample doesn't work, I'll try to get a hold of a Windows machine somewhere to see if I can reproduce it.

**Comment by doldman** [comment]
Hi Koen,
Yes, I can confirm that it has the same error.
return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 31946: character maps to <undefined>
I tried some variations of the magic encoding line (https://www.python.org/dev/peps/pep-0263/) but no luck.
Thanks,
Dominic


     On Monday, 22 June 2015, 23:19, Koen Van Daele <notifications@github.com> wrote:
   

 I've added a sample script: https://github.com/koenedaele/skosprovider/blob/master/examples/api.pyCould you try running this?The problem you're running into is probably Windows specific and has to do with character encodings. The sample I posted has a special header to tell python the data in the file is UTF-8. But I thought that was only needed for python 2.x and you seem to be running 3.4.If you can confirm the sample doesn't work, I'll try to get a hold of a Windows machine somewhere to see if I can reproduce it.—

**Comment by doldman** [comment]
Hi Koen
additional..

sys.getdefaultencoding() says that encoding is utf-8. I understand that this is the default for windows. 

There seems to be a lot of material about Python encoding on the web....

D

**Comment by koenedaele** [comment]
What editor are you using? It's not setting a different encoding on The file you're editing?

Oh, how I love character encoding issues... And not just in Python too.

**Comment by koenedaele** [comment]
By the way, if you just want to run the example, change la châtaigne to something else.

**Comment by doldman** [comment]
Hi Koen,
I am using PyCharm professional - but I am a Python learner not a professional! :-) 
D

 


     On Tuesday, 23 June 2015, 8:02, Koen Van Daele <notifications@github.com> wrote:
   

 By the way, if you just want to run the example, change la châtaigne to something else. —

**Comment by koenedaele** [comment]
This outlines a few things about working with encodings in PyCharm: https://www.jetbrains.com/pycharm/help/configuring-individual-file-encoding.html

Can you check if the editor isn't creating the files with a windows encoding?

Other than that, if you clone the repository and run the example from the repo directly, what does that do? eg.

 * git clone https://github.com/koenedaele/skosprovider.git
 * create some kind of virtualenv (with pyvenv or so)
 * go to the directory where you cloned skosprovider
 * pip install -r requirements.txt (optional in this case)
 * python setup.py develop
 * python examples/api.py

**Comment by doldman** [comment]
Yes, I have been messing around with this.
I changed everything to utf-8 and in any event it recognised that I had encoded it as utf-8 in the file and I think this overides settings. In any event even with utf-8 set for everything it still has the error.
I will clone now and set up a virtualenv now.
D
 


     On Tuesday, 23 June 2015, 16:26, Koen Van Daele <notifications@github.com> wrote:
   

 This outlines a few things about working with encodings in PyCharm: https://www.jetbrains.com/pycharm/help/configuring-individual-file-encoding.htmlCan you check if the editor isn't creating the files with a windows encoding?Other than that, if you clone the repository and run the example from the repo directly, what does that do? eg.   
   - git clone https://github.com/koenedaele/skosprovider.git
   - create some kind of virtualenv (with pyvenv or so)
   - go to the directory where you cloned skosprovider
   - pip install -r requirements.txt (optional in this case)
   - python setup.py develop
   - python examples/api.py

**Comment by koenedaele** [comment]
Damned, might really have to go and find a windows machine...

Serves me right for including French in a sample file I guess...

**Comment by doldman** [comment]
Hi Koen,
I get the same error in the virtual environment. Sorry!
D
 


     On Tuesday, 23 June 2015, 16:37, Koen Van Daele <notifications@github.com> wrote:
   

 Damned, might really have to go and find a windows machine...Serves me right for including French in a sample file I guess...—

**Comment by doldman** [comment]
Hi Koen,
Here is the error from the virtual environment.
D

(ENV) D:\gitrepository\skosprovider> python examples/api.py
Traceback (most recent call last):
  File "examples/api.py", line 7, in <module>
    from skosprovider.providers import DictionaryProvider
  File "d:\gitrepository\skosprovider\skosprovider\providers.py", line 21, in <m
odule>
    from .skos import (
  File "d:\gitrepository\skosprovider\skosprovider\skos.py", line 15, in <module
>
    from language_tags import tags
  File "D:\skosp\ENV\lib\site-packages\language_tags\__init__.py", line 1, in <m
odule>
    from language_tags.tags import tags
  File "D:\skosp\ENV\lib\site-packages\language_tags\tags.py", line 5, in <modul
e>
    from language_tags.Subtag import Subtag
  File "D:\skosp\ENV\lib\site-packages\language_tags\Subtag.py", line 8, in <mod
ule>
    registry = json.load(open(os.path.join(parent_dir, "data/json/registry.json"
)))
  File "C:\Python34\Lib\json\__init__.py", line 265, in load
    return loads(fp.read(),
  File "D:\skosp\ENV\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 31946: ch
aracter maps to <undefined> 
 



     On Tuesday, 23 June 2015, 16:45, Dominic Oldman <doint@oldman.me.uk> wrote:
   

 Hi Koen,
I get the same error in the virtual environment. Sorry!
D
 


     On Tuesday, 23 June 2015, 16:37, Koen Van Daele <notifications@github.com> wrote:
   

 Damned, might really have to go and find a windows machine...Serves me right for including French in a sample file I guess...—

**Comment by koenedaele** [comment]
Thanks Dominic. That confirms my suspicions. I was starting to suspect The problem wasn't in skosprovider at all, but in language tags. I'll try to fix it later tonight.

**Comment by koenedaele** [comment]
Dominic,

could you update the git clone you have (ie. git pull). Then update the dependencies (pip install -r requirements-dev.txt) and then run the example again. I think I've fixed the error in language-tags, but I want to be be sure before I release an update.

Fingers crossed,
Koen

**Comment by koenedaele** [comment]
When running the pip command, you should see something like:

```
Obtaining language-tags from git+https://github.com/OnroerendErfgoed/language-tags.git#egg=language-tags (from -r requirements.txt (line 2))
  Cloning https://github.com/OnroerendErfgoed/language-tags.git to /home/koen/Virtualenvs/skosprovider/src/language-tags
```

**Comment by doldman** [comment]
Hi Koen,

Many thanks!  This seems to work. I have tried in a virtual env and updated my main python libraries and it works in Pycharm also. 

I had decided to learn more Python using a skoprovider application project and got this problem straight away (I work at the British Museum where I have published the collection as CIDOC CRM linked data). I have dumps on a link off collection.britishmuseum.org (sparql endpoint - soon to be updated and made more robust)

For future reference it would be great to know the reason for this error. There is a huge amount of material on the web on this subject. 

Cheers,

Dominic

**Comment by koenedaele** [comment]
Hi Dominic,

we should be releasing a new version of the language-tags library later today. Thank you for your feedback and investing the time to get things sorted out. Much appreciated.

There's a bit more information on the error in OnroerendErfgoed/language-tags#7 Basically, we didn't tell python as what the file was encoded, so it had to guess. And for some reason on your system it guesses `cp-1252`, which is a very windows specific character set. And a few codepoints are undefined in it. So it can't decode the file properly and throws an error. I believe had it guessed something else, things might have worked but you might have also seen some strange characters.
You're probably the first Windows user we've had. Over here everything's run on Linux (both servers and developers machines), where I guess it guesses `utf-8` correctly.

I hadn't seen that BM website. I do know https://github.com/findsorguk/bmThesauri exists. Do they contain the same data? 
(BTW. I think we were both at the Nachos event at the BM in 2012 and I saw your presentation at CIDOC 2014 Dresden).

Anyway, thanks for the bug report. Let me know if any further issues arrise.

Cheers,
Koen

**Comment by doldman** [comment]
Hi Koen,


Yes, I started debugging and could see that at one point it defaulted to cp -1252.


Findsorguk is the finds database. Collection.British museum.org carries the 2 million or so object records for the BM. The sparql end point is a bit fkaky at the moment and we are about to change the software., but there is a tar dump also.


I will start building my alignment software now this is fixed.


D


Sent from Yahoo Mail on Android

**Comment by koenedaele** [comment]
The issues has been fixed in language-tags 0.3.2. I'll be releasing a version of new minor version of skosprovider with a requirements file that points to the new version.
